### PR TITLE
Added Sky Drop to Battle Engine

### PR DESCRIPTION
**Pull request by WiserVisor**
Sky Drop works just as it does from Gen 6 onwards.

## Description
When the target of Sky Drop is released, the text: "{TARGET} was released from Sky Drop!" doesn't display. I found that text to be unnecessary, since you can see the target's sprite reappear, and a bit annoying, so I left that out.

## **Pokecommunity contact info**
https://www.pokecommunity.com/member.php?u=682852

### CODE REVIEW

File: src/battle_util.c [hunk 1]
```diff
@@ -1420,12 +1421,123 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
-    gStatuses3[battler] &= ~(STATUS3_SEMI_INVULNERABLE);
+    // Don't clear battler's semi-invulnerable bits if they are held by Sky Drop.
+    if (gBattleStruct->skyDropTargets[1] != battler && gBattleStruct->skyDropTargets[3] != battler)
+        gStatuses3[battler] &= ~(STATUS3_SEMI_INVULNERABLE);
+    
+    // Check to see if this Pokemon was in the middle of using Sky Drop. If so, release the target.
+    if (gBattleStruct->skyDropTargets[0] == battler)
+    {
+        // Sets skyDropTargets[1] to be the battler id for the target
+        for(i = 0; i < MAX_BATTLERS_COUNT; i++)
```
Comment (LOuroboros) [inline src/battle_util.c:1438]:
```suggestion
        for (i = 0; i < MAX_BATTLERS_COUNT; i++)
```

File: src/battle_util.c [hunk 2]
```diff
@@ -1420,12 +1421,123 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+        {
+            if(gBattleStruct->skyDropTargets[1] == i)
```
Comment (LOuroboros) [inline src/battle_util.c:1440]:
```suggestion
            if (gBattleStruct->skyDropTargets[1] == i)
```

File: src/battle_util.c [hunk 3]
```diff
@@ -1420,12 +1421,123 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+                            gBattleStruct->skyDropTargets[0] = 0xFE - battler;
+
+                        // If this CancelMultiTurnMoves is occuring due to cancelmultiturnmoves script
+                        else if (gBattlescriptCurrInstr[0] == 0x76 && gBattlescriptCurrInstr[2] == 0)
+                        {
+                            gBattlerAttacker = i;
+                            gBattlescriptCurrInstr = BattleScript_ThrashConfuses - 3;
+                        }
+						
+						// If this CancelMultiTurnMoves is occuring due to receiving Sleep/Freeze status
```
Comment (LOuroboros) [inline src/battle_util.c:1475]: Spaces instead of tabs.

File: src/battle_util.c [hunk 4]
```diff
@@ -1420,12 +1421,123 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+        // Clear skyDropTargets data, unless this CancelMultiTurnMoves is caused by Yawn, attackcanceler, or VARIOUS_GRAVITY_ON_AIRBORNE_MONS
+        if (!(gBattleMons[gBattleStruct->skyDropTargets[1]].status2 & STATUS2_LOCK_CONFUSE) && gBattleStruct->skyDropTargets[0] < 4)
+        {
+            gBattleStruct->skyDropTargets[0] = 0xFF;
+            gBattleStruct->skyDropTargets[1] = 0xFF;
+        }
+    }
+    else if (gBattleStruct->skyDropTargets[2] == battler)
+    {
+        for(i = 0; i < MAX_BATTLERS_COUNT; i++)
```
Comment (LOuroboros) [inline src/battle_util.c:1498]:
```suggestion
        for (i = 0; i < MAX_BATTLERS_COUNT; i++)
```

File: src/battle_util.c [hunk 5]
```diff
@@ -1420,12 +1421,123 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+        {
+            if(gBattleStruct->skyDropTargets[3] == i)
```
Comment (LOuroboros) [inline src/battle_util.c:1500]:
```suggestion
            if (gBattleStruct->skyDropTargets[3] == i)
```

File: src/battle_util.c [hunk 6]
```diff
@@ -1420,12 +1421,123 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+                {
+                    gBattleMons[i].status2 &= ~(STATUS2_LOCK_CONFUSE);
+                    if (!(GetBattlerAbility(i) == ABILITY_OWN_TEMPO
+                        || gBattleMons[i].status2 & STATUS2_CONFUSION
+                        || IsBattlerTerrainAffected(i, STATUS_FIELD_MISTY_TERRAIN)))
+                    {
+                        gBattleMons[i].status2 |= STATUS2_CONFUSION_TURN(((Random()) % 4) + 2);
+                        
+                        if (gBattlescriptCurrInstr[0] == 0x0 || (gBattlescriptCurrInstr[0] == 0x76 && gBattlescriptCurrInstr[2] == 76))
+                            gBattleStruct->skyDropTargets[2] = 0xFE - battler;
```
Comment (LOuroboros) [inline src/battle_util.c:1517]:
```suggestion
                        if (gBattlescriptCurrInstr[0] == 0x0 || (gBattlescriptCurrInstr[0] == 0x76 && gBattlescriptCurrInstr[2] == 76))
                        {
                            gBattleStruct->skyDropTargets[2] = 0xFE - battler;
                        }
```

File: src/battle_util.c [hunk 7]
```diff
@@ -3438,7 +3562,7 @@ u8 AtkCanceller_UnableToUseMove(void)
             if ((gBattleMons[gBattlerAttacker].status1 & STATUS1_PARALYSIS) && (Random() % 4) == 0)
             {
                 gProtectStructs[gBattlerAttacker].prlzImmobility = TRUE;
-                // This is removed in Emerald for some reason
+                // This is removed in Emerald because it is called in BattleScript_MoveUsedIsParalyzed.
```
Comment (LOuroboros) [inline src/battle_util.c:3565]: You should revert this change and submit it to Pret's Pokeemerald repository instead.

File: src/battle_util.c [hunk 8]
```diff
@@ -9128,6 +9256,10 @@ bool32 CanMegaEvolve(u8 battlerId)
             && (mega->alreadyEvolved[partnerPosition] || (mega->toEvolve & gBitTable[BATTLE_PARTNER(battlerId)])))
             return FALSE;
     }
+	
+	// Check if mon is currently held by Sky Drop
+	if (gStatuses3[battlerId] & STATUS3_SKY_DROPPED)
+		return FALSE;
```
Comment (LOuroboros) [inline src/battle_util.c:9262]: Spaces instead of tabs.

**Review by LOuroboros — CHANGES_REQUESTED** [review]

File: src/battle_util.c [hunk 9]
```diff
@@ -1463,7 +1463,9 @@ void CancelMultiTurnMoves(u8 battler)
                         
                         // If this CancelMultiTurnMoves is occuring due to attackcanceller or VARIOUS_GRAVITY_ON_AIRBORNE_MONS
                         if (gBattlescriptCurrInstr[0] == 0x0 || (gBattlescriptCurrInstr[0] == 0x76 && gBattlescriptCurrInstr[2] == 76))
+                        {
                             gBattleStruct->skyDropTargets[0] = 0xFE - battler;
+                        }
```
Comment (LOuroboros) [inline src/battle_util.c]: I can't select all the lines involved, but the `else if` statements below this `if` should all come one after the other, that is without empty spaces inbetween, in order to keep consistency with upstream *(Pret)*.

**Review by LOuroboros — CHANGES_REQUESTED** [review]

File: src/battle_script_commands.c [hunk 10]
```diff
@@ -1609,22 +1609,23 @@ static bool32 AccuracyCalcHelper(u16 move)
         JumpIfMoveFailed(7, move);
         return TRUE;
     }
-    else if (GetBattlerAbility(gBattlerAttacker) == ABILITY_NO_GUARD)
+	// If the attacker has the ability No Guard and they aren't targeting a Pokemon involved in a Sky Drop with the move Sky Drop, move hits.
```
Comment (LOuroboros) [inline src/battle_script_commands.c:1612]: Spaces instead of tab.

File: src/battle_script_commands.c [hunk 11]
```diff
@@ -1609,22 +1609,23 @@ static bool32 AccuracyCalcHelper(u16 move)
...
+    else if (GetBattlerAbility(gBattlerAttacker) == ABILITY_NO_GUARD && (move != MOVE_SKY_DROP || (gBattlerTarget != gBattleStruct->skyDropTargets[0] && gBattlerTarget != gBattleStruct->skyDropTargets[1] && gBattlerTarget != gBattleStruct->skyDropTargets[2] && gBattlerTarget != gBattleStruct->skyDropTargets[3])))
     {
         if (!JumpIfMoveFailed(7, move))
             RecordAbilityBattle(gBattlerAttacker, ABILITY_NO_GUARD);
         return TRUE;
     }
-    else if (GetBattlerAbility(gBattlerTarget) == ABILITY_NO_GUARD)
+	// If the target has the ability No Guard and they aren't involved in a Sky Drop or the current move isn't Sky Drop, move hits.
```
Comment (LOuroboros) [inline src/battle_script_commands.c:1619]: Spaces instead of tab.

File: src/battle_script_commands.c [hunk 12]
```diff
@@ -2929,9 +2930,20 @@ void SetMoveEffect(bool32 primary, u32 certain)
                 else
                 {
                     gBattleMons[gEffectBattler].status2 |= STATUS2_CONFUSION_TURN(((Random()) % 4) + 2); // 2-5 turns
-
-                    BattleScriptPush(gBattlescriptCurrInstr + 1);
-                    gBattlescriptCurrInstr = sMoveEffectBS_Ptrs[gBattleScripting.moveEffect];
+                    
+					// If the confusion is activating due to being released from Sky Drop, go to "confused due to fatigue" script.
+					// Otherwise, do normal confusion script.
```
Comment (LOuroboros) [inline src/battle_script_commands.c:2935]: Spaces instead of tabs.

File: src/battle_script_commands.c [hunk 13]
```diff
@@ -8959,6 +8994,101 @@ static void Cmd_various(void)
...
+            gBattleStruct->skyDropTargets[3] = 0xFF;
+        }
+        
+        // Confuse target if they were in the middle of Petal Dance/Outrage/Thrash when targeted.
+        if (gBattleMons[gBattlerTarget].status2 & STATUS2_LOCK_CONFUSE)
+            gBattleScripting.moveEffect = (MOVE_EFFECT_CONFUSION | MOVE_EFFECT_CERTAIN);
+        return;
+    case VARIOUS_SKY_DROP_YAWN: // If the mon that's sleeping due to Yawn was holding a Pokemon in Sky Drop, release the target and clear Sky Drop data.
+        if (gEffectBattler == gBattleStruct->skyDropTargets[0])
+        {
```
Comment (LOuroboros) [inline src/battle_script_commands.c:9064]: There's 4 empty spaces after this bracket.

File: src/battle_util.c [hunk 14]
```diff
@@ -1420,12 +1421,127 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+                            BattleScriptPush(gBattlescriptCurrInstr + 1);
+                            gBattlescriptCurrInstr = BattleScript_ThrashConfuses - 1;
+                        }
+                    }
+                    
+                }
+                break;
+            }
+        }
+
```
Comment (LOuroboros) [inline src/battle_util.c:1490]:
```suggestion

```

File: src/battle_util.c [hunk 15]
```diff
@@ -1420,12 +1421,127 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+        }
+    }
+    else if (gBattleStruct->skyDropTargets[2] == battler)
+    {
+        for (i = 0; i < MAX_BATTLERS_COUNT; i++)
+        {
+            if (gBattleStruct->skyDropTargets[3] == i)
+            {
+                gStatuses3[i] &= ~(STATUS3_SKY_DROPPED | STATUS3_ON_AIR);
+
```
Comment (LOuroboros) [inline src/battle_util.c:1505]:
```suggestion

```

File: src/battle_util.c [hunk 16]
```diff
@@ -1420,12 +1421,127 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+    {
+        for (i = 0; i < MAX_BATTLERS_COUNT; i++)
+        {
+            if (gBattleStruct->skyDropTargets[3] == i)
+            {
+                gStatuses3[i] &= ~(STATUS3_SKY_DROPPED | STATUS3_ON_AIR);
+                
+                gSprites[gBattlerSpriteIds[battler]].invisible = FALSE;
+                gSprites[gBattlerSpriteIds[i]].invisible = FALSE;
+
```
Comment (LOuroboros) [inline src/battle_util.c:1508]:
```suggestion

```

File: src/battle_util.c [hunk 17]
```diff
@@ -1420,12 +1421,127 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+                            gBattleStruct->skyDropTargets[0] = 0xFE - battler;
+                        }
+
+                        // If this CancelMultiTurnMoves is occuring due to cancelmultiturnmoves script
+                        else if (gBattlescriptCurrInstr[0] == 0x76 && gBattlescriptCurrInstr[2] == 0)
+                        {
+                            gBattlerAttacker = i;
+                            gBattlescriptCurrInstr = BattleScript_ThrashConfuses - 3;
+                        }
+
```
Comment (LOuroboros) [inline src/battle_util.c:1476]:
```suggestion

```

File: src/battle_util.c [hunk 18]
```diff
@@ -1420,12 +1421,127 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+						
+                        // If this CancelMultiTurnMoves is occuring due to receiving Sleep/Freeze status
+                        else if (gBattleScripting.moveEffect <= PRIMARY_STATUS_MOVE_EFFECT)
+                        {
+                            gBattlerAttacker = i;
+                            BattleScriptPush(gBattlescriptCurrInstr + 1);
+                            gBattlescriptCurrInstr = BattleScript_ThrashConfuses - 1;
+                        }
+                    }
+
```
Comment (LOuroboros) [inline src/battle_util.c:1485]:
```suggestion

```

File: src/battle_util.c [hunk 19]
```diff
@@ -1420,12 +1421,127 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+                            gBattlerAttacker = i;
+                            BattleScriptPush(gBattlescriptCurrInstr + 1);
+                            gBattlescriptCurrInstr = BattleScript_ThrashConfuses - 1;
+                        }
+                    }
+                }
+                break;
+            }
+        }
+
```
Comment (LOuroboros) [inline src/battle_util.c:1538]:
```suggestion

```

File: src/battle_util.c [hunk 20]
```diff
@@ -1420,12 +1421,127 @@ void MarkBattlerReceivedLinkData(u8 battlerId)
...
+                
+                if (gBattleMons[i].status2 & STATUS2_LOCK_CONFUSE && gBattleStruct->turnEffectsTracker != 24)
+                {
+                    gBattleMons[i].status2 &= ~(STATUS2_LOCK_CONFUSE);
+                    if (!(GetBattlerAbility(i) == ABILITY_OWN_TEMPO
+                        || gBattleMons[i].status2 & STATUS2_CONFUSION
+                        || IsBattlerTerrainAffected(i, STATUS_FIELD_MISTY_TERRAIN)))
+                    {
+                        gBattleMons[i].status2 |= STATUS2_CONFUSION_TURN(((Random()) % 4) + 2);
+
```
Comment (LOuroboros) [inline src/battle_util.c:1517]:
```suggestion

```

File: src/battle_util.c [hunk 21]
```diff
@@ -9128,6 +9260,10 @@ bool32 CanMegaEvolve(u8 battlerId)
             && (mega->alreadyEvolved[partnerPosition] || (mega->toEvolve & gBitTable[BATTLE_PARTNER(battlerId)])))
             return FALSE;
     }
+
```
Comment (LOuroboros) [inline src/battle_util.c:9263]:
```suggestion

```

**Review by LOuroboros — CHANGES_REQUESTED** [review]

File: src/battle_util.c (hunk 21 above)
Comment (WiserVisor) [reply src/battle_util.c:9263]: I have to ask, why do you want all of these empty lines to be removed? I could somewhat understand it if that was how the rest of the code was formatted in the repository, (even though it can sometimes cause the code to be too close together and difficult to read) but that isn't the case here. In CanMegaEvolve, there are other if-statements that have empty lines after them, such as `if (holdEffect == HOLD_EFFECT_MEGA_STONE)` and `if (GetMegaEvolutionSpecies(species, itemId) != SPECIES_NONE)`, which help denote that one if-statement is not directly related to those that follow. There are other if-statements that don't have this extra line, admittedly, but there doesn't seem to be a consistent rule one way or the other. I'm still fine with making these changes if that's what it takes for the pull request to be accepted, but I'm curious to hear the reasoning for these changes, since they seem unnecessary to me.

Comment (LOuroboros) [reply src/battle_util.c:9263]: > I could somewhat understand it if that was how the rest of the code was formatted in the repository

I don't want the lines to be removed per se, just the spaces inside of them, and effectively, it's mainly to remain consistent with the rest of the code.

Comment (WiserVisor) [reply src/battle_util.c:9263]: Ohhhhhhh, I see. I misinterpreted your suggested changes. My bad!

File: src/battle_script_commands.c [hunk 22]
```diff
@@ -8959,6 +8994,101 @@ static void Cmd_various(void)
...
+            gBattlescriptCurrInstr = T1_READ_PTR(gBattlescriptCurrInstr + 3);
+        else
+            gBattlescriptCurrInstr += 7;
+        return;
+    case VARIOUS_SET_SKY_DROP:
+        gStatuses3[gBattlerTarget] |= (STATUS3_SKY_DROPPED | STATUS3_ON_AIR);
+        /* skyDropTargets holds the information of who is the attacker and the target of Sky Drop. 
+           This is needed in the case that multiple Pokemon use Sky Drop in the same turn or if
+           the target of a Sky Drop faints while in the air.*/
+        if (gBattleStruct->skyDropTargets[0] == 0xFF)
```
Comment (ghoulslash) [inline src/battle_script_commands.c:9009]: I'm confused about these `skyDropTargets` indices. are they not the battler positions? shouldnt it be something long the lines of:
```
gBattleStruct->skyDropTargets[gBattlerTarget] = gBattlerAttacker;
gBattleStruct->skyDropTargets[gBattlerAttacker] = gBattlerTarget;
```

File: src/battle_main.c [hunk 23]
```diff
@@ -3081,6 +3087,56 @@ void FaintClearSetData(void)
...
         if ((gBattleMons[i].status2 & STATUS2_WRAPPED) && *(gBattleStruct->wrappedBy + i) == gActiveBattler)
             gBattleMons[i].status2 &= ~(STATUS2_WRAPPED);
+
+        // If the fainted mon was holding another Pokemon in Sky Drop, release the mon and clear Sky Drop data
+        if (gActiveBattler == gBattleStruct->skyDropTargets[0] && i == gBattleStruct->skyDropTargets[1])
+        {
+            gBattleStruct->skyDropTargets[0] = 0xFF;
+            gBattleStruct->skyDropTargets[1] = 0xFF;
+            
+            gStatuses3[i] &= ~(STATUS3_SKY_DROPPED | STATUS3_ON_AIR);
```
Comment (ghoulslash) [inline src/battle_main.c:3097]: the status3 and status2 flags are cleared at the top of this function

File: src/battle_main.c [hunk 24]
```diff
@@ -3081,6 +3087,56 @@ void FaintClearSetData(void)
...
+            gSprites[gBattlerSpriteIds[i]].invisible = FALSE;
+            
+            // If the target was sky dropped in the middle of using Outrage/Petal Dance/Thrash,
+            // confuse them upon release and print "confused via fatigue" message and animation.
+            if (gBattleMons[i].status2 & STATUS2_LOCK_CONFUSE)
```
Comment (ghoulslash) [inline src/battle_main.c:3102]: the status3 and status2 flags are cleared at the top of this function, so this will always be false.

File: src/battle_main.c [hunk 25]
```diff
@@ -3081,6 +3087,56 @@ void FaintClearSetData(void)
             gBattleMons[i].status2 &= ~(STATUS2_INFATUATED_WITH(gActiveBattler));
         if ((gBattleMons[i].status2 & STATUS2_WRAPPED) && *(gBattleStruct->wrappedBy + i) == gActiveBattler)
             gBattleMons[i].status2 &= ~(STATUS2_WRAPPED);
+
+        // If the fainted mon was holding another Pokemon in Sky Drop, release the mon and clear Sky Drop data
+        if (gActiveBattler == gBattleStruct->skyDropTargets[0] && i == gBattleStruct->skyDropTargets[1])
```
Comment (ghoulslash) [inline src/battle_main.c:3092]: based on my other comment about `BattleStruct->skyDropTargets`, you should be able to just get the target to release from `BattleStruct->skyDropTargets[gActiveBattler]`

File: src/battle_main.c [hunk 26]
```diff
@@ -3081,6 +3087,56 @@ void FaintClearSetData(void)
...
+        }
+        else if (gActiveBattler == gBattleStruct->skyDropTargets[2] && i == gBattleStruct->skyDropTargets[3])
+        {
+            gBattleStruct->skyDropTargets[2] = 0xFF;
+            gBattleStruct->skyDropTargets[3] = 0xFF;
+            
+            gStatuses3[i] &= ~(STATUS3_SKY_DROPPED | STATUS3_ON_AIR);
+            gSprites[gBattlerSpriteIds[i]].invisible = FALSE;
+            
+            if (gBattleMons[i].status2 & STATUS2_LOCK_CONFUSE)
```
Comment (ghoulslash) [inline src/battle_main.c:3126]: same comment as above

File: src/battle_main.c [hunk 27]
```diff
@@ -3160,6 +3216,12 @@ void FaintClearSetData(void)
     UndoFormChange(gBattlerPartyIndexes[gActiveBattler], GET_BATTLER_SIDE(gActiveBattler), FALSE);
     if (GetBattlerSide(gActiveBattler) == B_SIDE_PLAYER)
         UndoMegaEvolution(gBattlerPartyIndexes[gActiveBattler]);
+    
+    // If the fainted Pokemon was being held by Sky Drop, clear their Sky Drop data.
+    if (gActiveBattler == gBattleStruct->skyDropTargets[1])
```
Comment (ghoulslash) [inline src/battle_main.c:3221]: should be able to just clear `gBattleStruct->skyDropTargets[gActiveBattler]` and `gBattleStruct->skyDropTargets[gBattleStruct->skyDropTargets[gActiveBattler]]`

File: src/battle_script_commands.c [hunk 28]
```diff
@@ -1609,22 +1609,23 @@ static bool32 AccuracyCalcHelper(u16 move)
         JumpIfMoveFailed(7, move);
         return TRUE;
     }
-    else if (GetBattlerAbility(gBattlerAttacker) == ABILITY_NO_GUARD)
+    // If the attacker has the ability No Guard and they aren't targeting a Pokemon involved in a Sky Drop with the move Sky Drop, move hits.
+    else if (GetBattlerAbility(gBattlerAttacker) == ABILITY_NO_GUARD && (move != MOVE_SKY_DROP || (gBattlerTarget != gBattleStruct->skyDropTargets[0] && gBattlerTarget != gBattleStruct->skyDropTargets[1] && gBattlerTarget != gBattleStruct->skyDropTargets[2] && gBattlerTarget != gBattleStruct->skyDropTargets[3])))
```
Comment (ghoulslash) [inline src/battle_script_commands.c:1613]: again, this logic should be heavily simplified by my `skyDropTargets` suggestion

File: src/battle_script_commands.c [hunk 29]
```diff
@@ -1609,22 +1609,23 @@ static bool32 AccuracyCalcHelper(u16 move)
...
     {
         if (!JumpIfMoveFailed(7, move))
             RecordAbilityBattle(gBattlerTarget, ABILITY_NO_GUARD);
         return TRUE;
     }
 
     if ((gStatuses3[gBattlerTarget] & STATUS3_PHANTOM_FORCE)
-        || (!(gBattleMoves[move].flags & FLAG_DMG_IN_AIR) && gStatuses3[gBattlerTarget] & STATUS3_ON_AIR)
-        || (!(gBattleMoves[move].flags & FLAG_DMG_2X_IN_AIR) && gStatuses3[gBattlerTarget] & STATUS3_ON_AIR)
+        || (!(gBattleMoves[move].flags & FLAG_DMG_IN_AIR || gBattleMoves[move].flags & FLAG_DMG_2X_IN_AIR) && gStatuses3[gBattlerTarget] & STATUS3_ON_AIR)
```
Comment (ghoulslash) [inline src/battle_script_commands.c:1628]: `|| (!(gBattleMoves[move].flags & (FLAG_DMG_IN_AIR | FLAG_DMG_2X_IN_AIR) && gStatuses3[gBattlerTarget] & STATUS3_ON_AIR)`

File: src/battle_script_commands.c [hunk 30]
```diff
@@ -7680,7 +7714,8 @@ static void Cmd_various(void)
         }
         return;
     case VARIOUS_GRAVITY_ON_AIRBORNE_MONS:
-        if (gStatuses3[gActiveBattler] & STATUS3_ON_AIR)
+        // Cancel all multiturn moves of IN_AIR Pokemon except those being targeted by Sky Drop.
+        if (gStatuses3[gActiveBattler] & STATUS3_ON_AIR && !(gActiveBattler == gBattleStruct->skyDropTargets[1] || gActiveBattler == gBattleStruct->skyDropTargets[3]))
```
Comment (ghoulslash) [inline src/battle_script_commands.c:7718]: should be able to just check status3 of `gActiveBattler` and gBattleStruct->skyDropTargets[gActiveBattler], if that field is not 0xFF

File: data/battle_scripts_1.s [hunk 31]
```diff
@@ -403,6 +403,65 @@ gBattleScriptsForMoveEffects::
 	.4byte BattleScript_EffectOctolock                @ EFFECT_OCTOLOCK
 	.4byte BattleScript_EffectClangorousSoul          @ EFFECT_CLANGOROUS_SOUL
 	.4byte BattleScript_EffectHit                     @ EFFECT_BOLT_BEAK
+	.4byte BattleScript_EffectSkyDrop                 @ EFFECT_SKY_DROP
+  
+BattleScript_EffectSkyDrop:
+	jumpifstatus2 BS_ATTACKER, STATUS2_MULTIPLETURNS, BattleScript_SkyDropTurn2
```
Comment (ghoulslash) [inline data/battle_scripts_1.s:409]: where is the check for if gravity is active?

**Review by ghoulslash — CHANGES_REQUESTED** [review]

File: include/constants/battle.h [hunk 32]
```diff
@@ -170,6 +170,7 @@
 #define STATUS3_AQUA_RING               (1 << 28)
 #define STATUS3_LASER_FOCUS             (1 << 29)
 #define STATUS3_POWER_TRICK             (1 << 30)
+#define STATUS3_SKY_DROPPED             (1 << 31) // Target of Sky Drop
 #define STATUS3_SEMI_INVULNERABLE       (STATUS3_UNDERGROUND | STATUS3_ON_AIR | STATUS3_UNDERWATER | STATUS3_PHANTOM_FORCE)
```
[the remark this answers is not in the document]
Comment (WiserVisor) [reply include/constants/battle.h:174]: Because the semi-invulnerable element of being held in Sky Drop is accounted for by `STATUS3_ON_AIR`, not `STATUS3_SKY_DROPPED`. The only advantage I see for adding `STATUS3_SKY_DROPPED` into `STATUS3_SEMI_INVULNERABLE` would be to simplify some of the clear-status statements, like `gStatuses3[i] &= ~(STATUS3_SKY_DROPPED | STATUS3_ON_AIR);` becoming `gStatuses3[i] &= ~(STATUS3_SEMI_INVULNERABLE);` in src/battle_util.c. If you think that's a necessary change to make, I will make it, but I didn't find it necessary as I was writing the code.

File: src/battle_script_commands.c (hunk 22 above)
Comment (WiserVisor) [reply src/battle_script_commands.c:9009]: They are not the battler positions. `skyDropTargets[0]` holds the battler position of the Pokémon that is currently holding another Pokémon in Sky Drop. This is the attacker of Sky Drop. That other Pokémon's battler position, a.k.a the target of a Sky Drop, is held in `skyDropTargets[1]`. The same can be said for indices 2 and 3, respectively, in case of a double battle where two instances of Sky Drop are occurring at the same time. This storage allows for the identification of both the attacker and the target of a particular instance of Sky Drop.

In short, the Pokémon at the battler position stored in index 0/2 is currently holding the Pokémon at the battler position stored in index 1/3 in Sky Drop, respectively.

File: src/battle_main.c (hunk 23 above)
Comment (WiserVisor) [reply src/battle_main.c:3097]: Status3 and status2 flags are only cleared for the fainted Pokémon (gActiveBattler), not for every Pokémon currently in battle. This means, if the attacker of a Sky Drop faints, it will be properly taken out of the semi-invulnerable state, but not the target. This section of code is to make sure the target is freed from Sky Drop and is removed from the in_air state.

Comment (ghoulslash) [reply src/battle_main.c:3097]: Good catch, thanks!

File: include/constants/battle.h (hunk 32 above)
[the remark this answers is not in the document]
Comment (ghoulslash) [reply include/constants/battle.h:174]: sounds good.

File: src/battle_script_commands.c (hunk 22 above)
Comment (ghoulslash) [reply src/battle_script_commands.c:9009]: This system is very convoluted. If you store the sky drop target as I suggested then you then have all the information you need to check the relevant sky drop pairs, even when two sky drops are simultaneously active, in a significantly more readable way.

e.g. consider the accuracy check. You can make it: `|| (!(gBattleMoves[move].flags & (FLAG_DMG_IN_AIR | FLAG_DMG_2X_IN_AIR)) && gStatuses3[gBattlerTarget] & STATUS3_ON_AIR
        && gBattleStruct->skyDropTarget[gBattlerAttacker] != gBattlerTarget)` instead of the long check you have now.

it also makes the set and clear sky drop various commands trivial.

File: asm/macros/battle_script.inc [hunk 33]
```diff
@@ -2055,3 +2055,25 @@
...
 
+	@ Will jump to script pointer if the target weighs less than 200 kg, or 441 lbs.
+	.macro jumpifunder200 battler:req, ptr:req
+	various \battler, VARIOUS_JUMP_IF_UNDER_200
+	.4byte \ptr
+	.endm
+
+	@ Sets the sky drop status and does all other necessary operations
+	.macro setskydrop
+	various 0, VARIOUS_SET_SKY_DROP
```
[the remark this answers is not in the document]
Comment (WiserVisor) [reply asm/macros/battle_script.inc:2066]: Aren't each of the changes in this file unnecessary, since gActiveBattler is not used in any of these functions?

File: src/battle_script_commands.c (hunk 28 above)
Comment (WiserVisor) [reply src/battle_script_commands.c:1613]: If a Pokémon attempts to use Sky Drop on another Pokémon that is involved in a different instance of Sky Drop, the move will always fail, regardless of accuracy and abilities. That's what the added conditions to the if-statement handles.

File: src/battle_script_commands.c (hunk 30 above)
Comment (WiserVisor) [reply src/battle_script_commands.c:7718]: Gravity will still affect both the target and the attacker, since the target is released and brought down once the attacker goes through CancelMultiTurnMoves. The added condition skips the unnecessary message "{target_of_sky_drop} can't stay airborne because of gravity!" The target's sprite will have already reappeared along with the attacker's, so there's no need for this message.

**Comment by AsparagusEduardo** [comment]
It has an error compiling after the BE merge

**Comment by AsparagusEduardo** [comment]
Finally had the time to fully test this again. There's just one more thing that needs fixing and I'm approving it.
Thank you so much for your patience, it's almost over ^^;
- [x] Target can't use moves.
- [x] Can't get hit except for moves that hit airborne targets.
- [x] Damages on the next turn.
- [x] Doesn't affect flying types.
- [x] Can target allies but fails.
- [x] No guard allows target being hit while lifted. (Unless the move used is also Sky Drop)
- [x] Lock-On/Mind Reader allows target being hit.
- [x] Fails if target is behind substitute
- [x] Fails if target is semi-invulnerable when starting move (including being Sky Dropped)
- [x] Cannot be used when Gravity is in effect
- [x] Is canceled if Gravity applies during semi-invulnerable turn
- [x] Discontinues moves like Rollout and Thrash. 
	- [x] For Thrash, it confuses when freed.
- [x] Protect guards against initial hit
- [x] Unaffected by Power Herb	
- [x] Not redirected by Follow Me.
	- [ ] If target with Follow Me is taken to the air, it no longer draw moves
- [x] Pokémon over 200kg can't be lifted (special message)
- [X] Can choose moves but doesn't have option to Mega Evolve or Ultra Burst (not yet implemented)
- [x] Wimp Out and Emergency Exit will not trigger if brought below half HP during semi-invulnerable turn.

**Review by AsparagusEduardo — APPROVED** [review]
Thank you, I'll now proceed to do the merge.
By the way, I don't think I've seen you in the RHH Discord server. Here's an invite to it, in case you ever need to ask or talk about the project: https://discord.gg/6CzjAG6GZk
